File: src/v2/Interleaving_String.java
/* Interleaving String
My Submissions
Given s1, s2, s3, find whether s3 is formed by the interleaving of s1 and s2.

For example,
Given:
s1 = "aabcc",
s2 = "dbbca",

When s3 = "aadbbcbcac", return true.
When s3 = "aadbbbaccc", return false.
*/


public class Interleaving_String {
    public boolean isInterleave(String s1, String s2, String s3) {
        if(s3.length() != s1.length() + s2.length())
            return false;
        if(s1.length() ==0 && s2.length() ==0 && s3.length() ==0)
            return true;
        int m = s1.length();
        int n = s2.length();
        boolean[][] array = new boolean[m+1][n+1];
        for(int i = 1 ; i<=m;i++){
            if(s1.substring(0,i).equals(s3.substring(0,i)))
                array[i][0] = true;
        }
        for(int i = 1 ; i<=n;i++){
            if(s2.substring(0,i).equals(s3.substring(0,i)))
                array[0][i] = true;
        }
        for(int i =1; i <=m;i++){
            for(int j =1; j<=n ;j++){
                if(array[i-1][j] && s1.charAt(i-1) == s3.charAt(i+j-1))
                    array[i][j] =true;
                if(array[i][j-1] && s2.charAt(j-1) == s3.charAt(i+j-1))
                    array[i][j] =true;
            }
        }
        return array[m][n];

    }

    public static void main(String[] args) {
        String s1= "";
        String s2 ="b";
        String s3 ="b";
        System.out.println(new Interleaving_String().isInterleave(s1, s2, s3));

    }

}

File: src/cpp/interleaving-string.cpp
#include <iostream>
#include <map>
using namespace std;
/***
For example,
Given:
s1 = "aabcc",
s2 = "dbbca",
When s3 = "aadbbcbcac", return true.
When s3 = "aadbbbaccc", return false.
*/
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) 
    {
    	if ( s3.size() == 0 ) return true;
    	if ( s3.size() != s1.size() + s2.size()) return false;
    	map<string,bool> dic;
    	//char tmp = s1[100];
        return innerIsInterleave(s1,s2,s3,dic);
        //return false;
    }
    bool innerIsInterleave(string s1, string s2, string s3,map<string,bool> &dic)
    {
    	//printf("%s,%s,%s\n",s3.c_str(),s1.c_str(),s2.c_str() );
    	if ( s3.size() == 0 ) return true;
    	int a = 0; int b = 0;
    	int index = 0;
    	bool result = false;
    	string key = s1+","+s2+","+s3;
    	if ( dic.find(key) != dic.end() )
    	{
    		result = dic[key];
    	}
    	else
    	{
    		for (int i = 0; i < s3.size(); ++i)
	    	{
	    		index++;
	    		char tmp = s3[i];
	    		//printf("s3[%d]=%c,s1[%d]=%c,s2[%d]=%c\n",i,s3[i],a,s1[a],b,s2[b]);
	    		if ( tmp == s1[a] &&  tmp == s2[b] )
	    		{
	    			result = innerIsInterleave(s1.substr(a+1),s2.substr(b),s3.substr(i+1),dic);
	    			if( result == true ) break;
	    			result = innerIsInterleave(s1.substr(a),s2.substr(b+1),s3.substr(i+1),dic);
	    			break;
	    		}
	    		else if( tmp == s1[a] ) a++;
	    		else if( tmp == s2[b] ) b++;
	    		else
	    		{
	    			result = false;
	    			break;
	    		}
	    	}
	    	//printf("a=%d,b=%d,%ld\n",a,b,s3.size() );
	    	if ( a + b == s3.size() )
	    	{
	    		result = true;
	    	}
	    	dic[key] = result;
    	}
    	return result;
    }
};
int main(int argc, char const *argv[])
{
	Solution s;
	//string s1 = "bbbbbabbbbabaababaaaabbababbaaabbabbaaabaaaaababbbababbbbbabbbbababbabaabababbbaabababababbbaaababaa";
	//string s2 = "babaaaabbababbbabbbbaabaabbaabbbbaabaaabaababaaaabaaabbaaabaaaabaabaabbbbbbbbbbbabaaabbababbabbabaab";
	//string s3 = "babbbabbbaaabbababbbbababaabbabaabaaabbbbabbbaaabbbaaaaabbbbaabbaaabababbaaaaaabababbababaababbababbbababbbbaaaabaabbabbaaaaabbabbaaaabbbaabaaabaababaababbaaabbbbbabbbbaabbabaabbbbabaaabbababbabbabbab";
	
	//string s1 = "aabcc";
	//string s2 = "dbbca";
	//string s3 = "aadbbcbcac";

	string s1 = "a";
	string s2 = "b";
	string s3 = "a";
	
	//string s4 = "aadbbbaccc";
	bool result1 = s.isInterleave(s1,s2,s3);
	bool result2 = false;//s.isInterleave("abbbbbbcabbacaacccababaabcccabcacbcaabbbacccaaaaaababbbacbb","ccaacabbacaccacababbbbabbcacccacccccaabaababacbbacabbbbabc",s4);
	printf("%s\n",result1 == true ? "true" : "false");
	return 0;
}